### ISSUE TITLE
Feature Request: potential improvements to pod install times in large projects

### ISSUE THREAD

**Issue by biboran** [CocoaPods/CocoaPods#9982]
🌈

* [x] I've read and understood the [*CONTRIBUTING* guidelines and have done my best effort to follow](https://github.com/CocoaPods/CocoaPods/blob/master/CONTRIBUTING.md).

# Report 

## What did you do?

Ran `pod install` on a big mono repository project with 100+ local pods and 100k+ files.

## What did you expect to happen?

Pods would be installed relatively quickly.

## What happened instead?

Pods installation would take up to 8 minutes even when all remote pods would be cached.

## CocoaPods Environment

<details>
  <summary>pod env</summary>
  <p>

### Stack

```
   CocoaPods : 1.9.3
        Ruby : ruby 2.4.1p111 (2017-03-22 revision 58053) [x86_64-darwin17]
    RubyGems : 2.6.14
        Host : Mac OS X 10.15.4 (19E224g)
       Xcode : 11.5 (11E608c)
         Git : git version 2.23.0
Ruby lib dir : /Users/user/.rvm/rubies/ruby-2.4.1/lib
Repositories : master - git - https://github.com/CocoaPods/Specs.git @ a8823ea555ec54359b3fbdc6efe6cd3b1ae15ff9

               trunk - CDN - https://cdn.cocoapods.org/
```

### Installation Source

```
Executable Path: /Users/user/.rvm/gems/ruby-2.4.1/bin/pod
```

### Plugins

```
claide-plugins          : 0.9.2
cocoapods-amimono       : 0.0.12
cocoapods-deintegrate   : 1.0.4
cocoapods-dependencies  : 1.0.0.beta.1
cocoapods-plugins       : 1.0.0
cocoapods-search        : 1.0.0
cocoapods-stats         : 1.1.0
cocoapods-trunk         : 1.4.1
cocoapods-try           : 1.1.0
cocoapods_debug         : 0.1.0
```
  </p>
</details>

## Project that demonstrates the issue

I have made a somewhat contrived [example](https://github.com/biboran/SlowCocoaPods). It is a CocoaPods project in the `Application` directory with 4 local pods and a huge `ci_scripts` directory. On my machine `pod install` runs in 13 seconds with the `ci_scripts` directory and in 1.5 seconds without the directory.

The issue in the example is quite evident. `SharedModels.podspec` and `SharedServices.podspec` search files by the glob pattern. To find the files which satisfy the pattern both pods have to go through the `ci_scripts` directory which takes some time. 

## Background

Our project has a bunch of `.podspec` files in the root directory for sharing code across other pods that contain code specific to the user stories. As a mono repository, it also has multiple other unrelated Xcode projects, some CI scripts, and a lot of other miscellaneous files. For a long time, such configuration worked just fine and the performance of `pod install` was good but as more and more files got added the performance slowly became miserable.

Unfortunately, it was not immediately obvious what the problem was. By profiling CocoaPods we found that [read_file_system](https://github.com/CocoaPods/CocoaPods/blob/ef91fa13b2c19fc1e84aa38e19e658e5cf84c212/lib/cocoapods/sandbox/path_list.rb#L50) was taking about 60% of the `pod install` time.

We made a quick-fix by monkey patching `read_file_system`:

<details>
<summary>quick_fix_read_file_system.rb</summary>
<p>

```ruby
require 'find'

# prepare static cache in pre_install
def prepare_file_cache
  $all_files = []
  $all_dirs = []

  repo_root_path = repo_root
  
  # ignore large irrelevant folders
  paths_to_ignore = %w[.git .idea ci_scripts some_other_irrelevant_folder].map { |path| repo_root_path + "/#{path}" }
  root_paths_to_cache = Dir.glob(repo_root_path + "/*").select { |path| File.directory?(path) && (paths_to_ignore.include?(path) == false) }
  root_paths_to_cache.each { |path| 
      Find.find(pod_path) do |path|
      (File.directory?(path) ? $all_dirs : $all_files) << path
    end
  }
end

module Pod
  class Sandbox
    class PathList
      def read_file_system
        unless root.exist?
          raise Informative, "Attempt to read non existent folder `#{root}`."
        end

        root_path_str = root.cleanpath.to_s + File::SEPARATOR
        root_length = root_path_str.length

        # Attempt to find dirs and files in the static cache
        dirs = ($all_dirs || []).select { |path| path.start_with?(root_path_str) }.map { |path| path.slice(root_length, path.length - root_length) } || []
        files = ($all_files || []).select { |path| path.start_with?(root_path_str) }.map { |path| path.slice(root_length, path.length - root_length) } || []

        # If none were found resort to reading from disk
        if dirs.empty? and files.empty?
          Find.find(root.to_s) do |path|
            directory = File.directory?(path)
            path = path.slice(root_length, path.length - root_length)
            next if path.nil?

            (directory ? dirs : files) << path
          end
        end

        dirs.sort_by!(&:upcase)
        files.sort_by!(&:upcase)

        @dirs = dirs
        @files = files
        @glob_cache = {}
      end
    end
  end
end
```
</p>
</details>

## Feature Request

As far as I understand `read_file_system` is called to cache all relevant files for each `PodTarget`(i.e. `.podspec` file) before any glob expressions are resolved. One solution might be to notify CocoaPods users that their local pods are misconfigured for performance. For example, CocoaPods could output a warning at the end of a "lengthy" `pod install`:

> [!] Local pod SharedModels resolves only 4% of the read files. Consider moving it to a deeper folder.

Another solution might be to create a cache for the local pods to prevent redundant fs reads.

Would such changes benefit CocoaPods users or does it sound to niche? Perhaps there are another ways to illustrate/fix this performance issue?

**Comment by dnkoutso** [comment]
Yes! This could be a great improvement. We use CocoaPods in a monorepo as well with hundreds of pods. Want to make a PR for it?

**Comment by biboran** [comment]
Sure, I'll start working on it!

**Comment by Usipov** [comment]
@dnkoutso, how much time does pod install take on your project?

**Comment by dnkoutso** [comment]
@Usipov 3 or 4 minutes but we also use incremental installation which we built for this specific reason and on average it can take 40-50 seconds instead.

**Comment by dnkoutso** [comment]
I may have misinterpreted or misread the initial report. I had thought we would be interested in actually improving the performance.

The warning is helpful but I think CocoaPods should instead fix its performance issue rather than try to gauge a folder structure teams should have.

**Comment by biboran** [comment]
That's a bummer. Guess I should've asked explicitly about which feature I should implement. Then I'll start working on the cache.

**Comment by dnkoutso** [comment]
@biboran I am sorry for the miscommunication from my side. If we can improve performance then that would be best in my opinion
